File: leetcode/math/296_Best_Meeting_Point/codes/math.py
from typing import List

class Solution:
    def minTotalDistance(self, grid: List[List[int]]) -> int:
        def get_coords():
            rows, cols = [], []
            for i in range(len(grid)):
                for j in range(len(grid[0])):
                    if grid[i][j] == 1:
                        rows.append(i)
            for j in range(len(grid[0])):
                for i in range(len(grid)):
                    if grid[i][j] == 1:
                        cols.append(j)
            return rows, cols

        def min_dist(coords):
            coords.sort()
            median = coords[len(coords)//2]
            return sum(abs(c - median) for c in coords)

        rows, cols = get_coords()
        return min_dist(rows) + min_dist(cols)

File: leetcode/math/296_Best_Meeting_Point/codes/math.c
#include <stdlib.h>

int cmp(const void* a, const void* b) {
    return (*(int*)a - *(int*)b);
}

int minTotalDistance(int** grid, int gridSize, int* gridColSize) {
    int rows[40000], cols[40000]; 
    int rCount = 0, cCount = 0;

    for (int i = 0; i < gridSize; i++)
        for (int j = 0; j < gridColSize[0]; j++)
            if (grid[i][j] == 1)
                rows[rCount++] = i;

    for (int j = 0; j < gridColSize[0]; j++)
        for (int i = 0; i < gridSize; i++)
            if (grid[i][j] == 1)
                cols[cCount++] = j;

    qsort(rows, rCount, sizeof(int), cmp);
    qsort(cols, cCount, sizeof(int), cmp);

    int row_median = rows[rCount / 2];
    int col_median = cols[cCount / 2];
    int sum = 0;

    for (int i = 0; i < rCount; i++) sum += abs(rows[i] - row_median);
    for (int i = 0; i < cCount; i++) sum += abs(cols[i] - col_median);

    return sum;
}
